Added New Test Definitions error, seekable, etc. Remove redundant test runs & Introduce dry test runs

nairashwin952013 opened the pull request:
# Goals
- Resolves #336 .
- Eliminate redundant integration-test runs.
- Test-Suite Size Reduction obtained from `609` tests -> `250` tests. 
- Class-Level Definition for every new integration-test that will be added in the future.
- Incorporate dry run capability for the TestNG Suite of SDK to determine test-methods that are executed and the order in which they are run, without actually executing them.
- These changes will further help in setting up a new Jenkins job `java-manta-pre-it` or simply update the existing one `java-manta-it` for new PR's made against the `Java SDK`.

# Challenges
Unfortunately TestNG doesn't comprehensively support filtering a test suite by groups so we **had** to be more explicit about suite definition.

# Overview
Adds a `it.dryRun` property that can be used to list the tests that would run with their respective TestNG params. TestNG unfortunately doesn't completely support filtering a test suite by groups so thorough detailing was required in suite definition.

Output of `mvn verify it.dryRun=true` using current `testng-it.xml` used with `TestListingInterceptor`:
```
[main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Base manta path for suite Java Manta SDK Integration Test Suite: /ashwin.nair/stor/java-manta-integration-tests/8669c9af-b512-4ac6-b7be-2390b6a5e61c/
[main] INFO  com.joyent.test.util.TestListingInterceptor [ ] - DRY-RUN: Listing [250] tests that would have run
--------------------------ALL-TESTS----------------------------------------------------------
main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 250
[main] WARN  com.joyent.test.util.MantaPathSuiteListener [ ] - Skipping suite cleanup since dry-run is enabled.
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.952 s - in TestSuite
```

List for the output for [before](https://gist.github.com/tjcelaya/fa82e54bd4be55ec95e1146954d7bd1f) & [after](https://gist.github.com/nairashwin952013/ae0a4799525654588bcce24e9595fd6a)  `-Dit.dryRun=true` respectively . Note that`dry-run-after` list is accurate since it shows single value for `encryptionCipher` i.e `AES128/CTR/NoPadding`. Other testing groups include `cbc-cipher` and `gcm-cipher` are also defined. 

# Changes

- change `usingEncryption` (boolean) to just `encryptionCipher` (string). omission (null) disables encryption

- add a new constructor for `IntegrationTestConfigContext` which handles the above

- switch from `@BeforeClass` with `@Parameters` to test constructors with just `@Parameters`. Whenever `SkipException` was thrown or `putDirectory` was called in a test which needed `@Parameters` a minimal method was created for `@BeforeClass`. The goal here was to make it safe to instantiate test classes without incurring side-effects.

- remove encryption param from tests which don't actually perform any object PUTs or MPU uploads (e.g. `MantaClientDirectoriesIT`)

- change `<packages>` to `<classes>` and target specific classes. 

- document a new system property `it.dryRun` which can be used by the new interceptor. When this system property is truthy as defined by `BooleanUtils#toBooleanObject`, i.e. the following are all true: `y`, `Y`, `t`, `T`, `on`, `ON`, `yes`, `YES`, `true`, `TRUE` and all the mixed-case versions thereof

- added a test listener called `TestListingInterceptor` which is used to collect suite definitions and disable actually running tests in the event `it.dryrun` is `true`.

- defined new `<test name></test>` and the **EXISTING** ones  in `testing-it.xml` more accurately using class-level specifications including **TestNG Group Definitions** : `timeout`, `range-downloads`, `snaplinks`, `seekable`, `error`, `headers`, `directory`, `metadata`, etc.

- Following up on #492 which **had** to be closed due to ```dependency-vulnerabilities```.

- Changes are incorporated from the closed PR #364 by Tomas which is to remove redundancies in ```integration-test``` runs.

cburroughs commented:
Drive by color opinion: "nightly" isn't a term exclusive to Joyent, but still pretty far on the jargon scale.  Is there a more self-descriptive term we could use? (slow? expensive?)

nairashwin952013 commented:
> Drive by color opinion: "nightly" isn't a term exclusive to Joyent, but still pretty far on the jargon scale. Is there a more self-descriptive term we could use? (slow? expensive?)

Expensive Sounds Good.

cburroughs commented:
I have not used testng groups before and am not sure if I understand how the flags are to be used.  With `-Dit.dryRun=true` I got `Expected test count: 232`. If I added `-Dgroups=headers` I got `Expected test count: 7`, which seemed okay.  But `-DexcludeGroups=expensive` and showed `Expected test count: 232`.

nairashwin952013 commented:
> I have not used testng groups before and am not sure if I understand how the flags are to be used. With `-Dit.dryRun=true` I got `Expected test count: 232`. If I added `-Dgroups=headers` I got `Expected test count: 7`, which seemed okay. But `-DexcludeGroups=expensive` and showed `Expected test count: 232`.

Its because the `expensive` test group is `enabled-false` by default

dekobon commented:
When running the test from Maven using `mvn verify`, I got the following errors:
```
[ERROR] Tests run: 221, Failures: 3, Errors: 0, Skipped: 25, Time elapsed: 2,381.243 s <<< FAILURE! - in TestSuite
[ERROR] beforeClass(com.joyent.manta.client.MantaClientMetadataIT)  Time elapsed: 0.019 s  <<< FAILURE!
java.lang.NullPointerException
	at com.joyent.manta.client.MantaClientMetadataIT.beforeClass(MantaClientMetadataIT.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:59)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:458)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:222)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:142)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:168)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

[ERROR] beforeClass(com.joyent.manta.client.MantaClientMetadataIT)  Time elapsed: 4.03 s  <<< FAILURE!
java.lang.NullPointerException
	at com.joyent.manta.client.MantaClientMetadataIT.beforeClass(MantaClientMetadataIT.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:59)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:458)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:222)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:142)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:168)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

[ERROR] beforeClass(com.joyent.manta.client.MantaClientMetadataIT)  Time elapsed: 9.977 s  <<< FAILURE!
java.lang.NullPointerException
	at com.joyent.manta.client.MantaClientMetadataIT.beforeClass(MantaClientMetadataIT.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:59)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:458)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:222)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:142)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:168)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:105)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR] com.joyent.manta.client.MantaClientMetadataIT.beforeClass(com.joyent.manta.client.MantaClientMetadataIT)
[ERROR]   Run 1: MantaClientMetadataIT.beforeClass:51 NullPointer
[INFO]   Run 2: PASS
[ERROR]   Run 3: MantaClientMetadataIT.beforeClass:51 NullPointer
[INFO]   Run 4: PASS
[ERROR]   Run 5: MantaClientMetadataIT.beforeClass:51 NullPointer
[INFO]   Run 6: PASS
[INFO] 
[INFO] 
[ERROR] Tests run: 205, Failures: 1, Errors: 0, Skipped: 11
```

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/manta/client/MantaClientMetadataIT.java:
```diff
@@ -26,24 +26,28 @@
-    @Parameters({"usingEncryption"})
-    public void beforeClass(@Optional Boolean usingEncryption) throws IOException {
+    @Parameters({"encryptionCipher"})
+    public void beforeClass(final @Optional String encryptionCipher) throws IOException {
```
Refactoring `beforeClass()` -> `MantaClientMetadataIT()` class constructor which was originally raising the `NullPointerException`

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/manta/client/MantaClientMetadataIT.java:
```diff
@@ -26,24 +26,28 @@
  * Tests for verifying the behavior of metadata with {@link MantaClient}.
  *
  * @author <a href="https://github.com/dekobon">Elijah Zupancic</a>
+ * @author <a href="https://github.com/nairashwin952013">Ashwin A Nair</a>
```
I will be removing the @author tags on all affected `integration-tests` within this PR now.

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedServerSideMultipartManagerSerializationIT.java:
```diff
@@ -36,40 +36,51 @@
+ * Tests for verifying the behavior of {@link com.joyent.manta.serialization.EncryptedMultipartSerializer}
+ * and {@link EncryptedServerSideMultipartManager} with {@link MantaClient}.
+ *
+ * @author <a href="https://github.com/dekobon">Elijah Zupancic</a>
```
Noted.

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedServerSideMultipartManagerIT.java:
```diff
@@ -47,29 +46,44 @@
+ * Tests for verifying the behavior of {@link EncryptedServerSideMultipartManager} with
+ * {@link MantaClient}.
+ *
+ * @author <a href="https://github.com/dekobon">Elijah Zupancic</a>
```
Noted.

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedJobsMultipartManagerIT.java:
```diff
@@ -55,7 +55,7 @@
  * is very long and expensive. */
 @Deprecated
-@Test(groups = { "encrypted" }, enabled = false)
+@Test(groups = { "expensive" }, enabled = false)
```
Your suggestion was factored into determining which `integration-tests` and `sub-tests` within other integration-tests will be defined as expensive (nightly). Also, I believe the reasoning behind your suggestion was because the class was deprecated and comprised of **heavy** operations.

nairashwin952013 replied:
The discussion was made [here](https://github.com/joyent/java-manta/pull/492#issuecomment-475365013).

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedServerSideMultipartManagerSerializationIT.java:
```diff
@@ -36,40 +36,51 @@
+ * @author <a href="https://github.com/dekobon">Elijah Zupancic</a>
+ * @author <a href="https://github.com/nairashwin952013">Ashwin A Nair</a>
+ */
+@Test(groups = {"encryptable", "multipart"})
```
Refactored groups that were initially defined as `encrypted` to `encryptable` so as to make a clear distinction between names of groups since we have a new test group defined as `unencrypted` [here](https://github.com/joyent/java-manta/pull/507/files#diff-dd127e7993afbd9ec23c3cfc2828d155R70) that comprises of all tests verifying operations in the SDK which **don't utilize** encryption.

dekobon replied:
My objection to the word "encryptable" is that the word sounds a bit unnatural. Why is it preferred over "encrypted"?

nairashwin952013 replied:
Will be refactoring `encryptable` -> `encrypted`

nairashwin952013 replied:
Review noted, will come up with a solution.

nairashwin952013 replied:
We might have to use `beanshell` scripts which will defined in the `testing-it.xml` file.

References: 
- [Test-NG Docs](https://testng.org/doc/documentation-main.html#beanshell).
- Blog [post](https://rationaleemotions.wordpress.com/2013/08/21/beanshell-and-testng/) that I found online.

nairashwin952013 commented:
> > I have not used testng groups before and am not sure if I understand how the flags are to be used. With `-Dit.dryRun=true` I got `Expected test count: 232`. If I added `-Dgroups=headers` I got `Expected test count: 7`, which seemed okay. But `-DexcludeGroups=expensive` and showed `Expected test count: 232`.
> 
> Its because the `expensive` test group is `enabled-false` by default



> I have not used testng groups before and am not sure if I understand how the flags are to be used. With `-Dit.dryRun=true` I got `Expected test count: 232`. If I added `-Dgroups=headers` I got `Expected test count: 7`, which seemed okay. But `-DexcludeGroups=expensive` and showed `Expected test count: 232`.

Also, I believe the comment [here](https://github.com/joyent/java-manta/pull/507/files#diff-dd127e7993afbd9ec23c3cfc2828d155R89) might be useful.

dekobon commented on java-manta-it/src/test/resources/testng-it.xml:
```diff
@@ -11,65 +12,190 @@
         </classes>
     </test>
-
+    <test name="Manta Client Directory Tests">
```
In the below changes, we moved from wildcard test definition to explicit test definition. This will make it really easy for engineers to forget to add a test to this file after they create it. Is there a way for us to define the tests more dynamically?

nairashwin952013 replied:
We might have to use `beanshell` scripts which will be defined in the `testing-it.xml` file.

References: 
- [Test-NG Docs](https://testng.org/doc/documentation-main.html#beanshell).
- Blog [post](https://rationaleemotions.wordpress.com/2013/08/21/beanshell-and-testng/) that I found online.

dekobon replied:
Let's leave it as is, but please add a section to the documentation calling out that the tests must be added to the XML files.

nairashwin952013 replied:
I think for Chris to run the `nightly` tests for the `Java SDK` in **Jenkins**. It would be preferable and much convenient if he is to run the test group from the command line , for eg 
`mvn clean test -DgroupToRun=nightly`
rather than figuring it out how to invoke them using `TestNG` classes.

nairashwin952013 replied:
> Let's leave it as is, but please add a section to the documentation calling out that the tests must be added to the XML files.

Testing Documentation updated with the latest [commit](https://github.com/joyent/java-manta/pull/507/commits/c3638ac33ddeafc330cd336c0252cc121fef8c19).

dekobon commented on java-manta-it/src/test/java/com/joyent/manta/client/multipart/EncryptedServerSideMultipartManagerIT.java:
```diff
@@ -47,29 +46,41 @@
+ * Tests for verifying the behavior of {@link EncryptedServerSideMultipartManager} with
+ * {@link MantaClient}.
+ */
+@Test(groups = {"encryptable", "multipart"})
```
My objection to the word "encryptable" is that the word sounds a bit unnatural. Why is it preferred over "encrypted"?

nairashwin952013 commented:
`Deprecated` classes coming within `expensive` testng-groups are being **disabled** by default

cburroughs commented:
I have -- I think -- the latest branch checked out.   Git tells me the last commit is 4d6d73bb4fff6b9aa477720b5bd496572ab0aa00.  Do I have that right?

When I do `mvn verify -Dit.dryRun=true` I get 
```
[main] INFO  com.joyent.test.util.TestListingInterceptor [ ] - DRY-RUN: Listing [294] tests that would have run
```

But when I run mvn verify -Dit.dryRun=true -DexcludedGroups=expensive
```
[main] INFO  com.joyent.test.util.TestListingInterceptor [ ] - DRY-RUN: Listing [322] tests that would have run
````

Do these counts match yours?  Shouldn't the number of expected tests go down when any `-DexcludedGroups` are given?

nairashwin952013 commented:
> I have -- I think -- the latest branch checked out. Git tells me the last commit is [4d6d73b](https://github.com/joyent/java-manta/commit/4d6d73bb4fff6b9aa477720b5bd496572ab0aa00). Do I have that right?
> 
> When I do `mvn verify -Dit.dryRun=true` I get
> 
> ```
> [main] INFO  com.joyent.test.util.TestListingInterceptor [ ] - DRY-RUN: Listing [294] tests that would have run
> ```
> 
> But when I run mvn verify -Dit.dryRun=true -DexcludedGroups=expensive
> 
> ```
> [main] INFO  com.joyent.test.util.TestListingInterceptor [ ] - DRY-RUN: Listing [322] tests that would have run
> ```
> 
> Do these counts match yours? Shouldn't the number of expected tests go down when any `-DexcludedGroups` are given?

Yes, These counts match mine, but when you invoke -`DexcludedGroups` parameter from the command line, it overrides all the exclusion tags defined in the `testng-it.xml` file which consecutively results in leading to more **redundant test** runs. This PR is made so that we won't have to invoke `-DexcludedGroups `anymore and eliminates redundant test runs since we can now source the `testng-it` (Integration-Test suite) to carry out the exclusions for us.

nairashwin952013 commented:
@cburroughs the discussion [here](https://github.com/joyent/java-manta/issues/336#issuecomment-331504606) will shed more light on what I am talking about.

cburroughs commented:
Glad to hear I'm not crazy about the counts.  Per our call, we have two use cases we want to be easy to from the command line (for jenkins).  Anything fancier is okay as a convenience for developers and their IDEs:
 (1) Run all of the tests we care about.  This would be for nightly and after a commit to master.
 (2) Run all of the tests we care about, *except* the really expensive ones.  This would be for testing PRs.

With this PR the answer to (1) is just the default from running `mvn verify`.  What is answer for (2)?

dekobon commented:
I just noticed that `EncryptedServerSideMultipartManagerIT` doesn't have `canListUploadsInProgress` marked as `@Test(groups = { "expensive" })`.

Additionally, when running `mvn clean verify` it doesn't skip tests in the `expensive` group by default. Is that behavior by design?

dekobon commented:
I just noticed that multi-threaded tests were enabled in the testng configuration. What testing / review have we done to ensure that the tests properly support being run in multiple threads?

dekobon commented:
Have you reviewed the output of `java-manta/java-manta-it/target/failsafe-reports/index.html` to verify that the tests actually run are the same as what you intended?

nairashwin952013 commented:
> Additionally, when running `mvn clean verify` it doesn't skip tests in the `expensive` group by default. Is that behavior by design?

Yes, I am aware of that, will come up with a work-around soon without affecting the testing suite, `TestNG` has limited resources in this regard.

nairashwin952013 commented:
> Have you reviewed the output of `java-manta/java-manta-it/target/failsafe-reports/index.html` to verify that the tests actually run are the same as what you intended?

Yes, since I incorporated `-Dit.dryRun=true`, it's been much easier to evaluate the tests that will be running while leveraging `mvn` CLI commands

nairashwin952013 commented:
> I just noticed that multi-threaded tests were enabled in the testng configuration. What testing / review have we done to ensure that the tests properly support being run in multiple threads?

Noted.

nairashwin952013 commented:
> I just noticed that `EncryptedServerSideMultipartManagerIT` doesn't have `canListUploadsInProgress` marked as `@Test(groups = { "expensive" })`.

The [discussion](https://github.com/joyent/java-manta/pull/492#issuecomment-475365013) here didn't include `EncryptedServerSideMultipartManagerIT.canListUploadsInProgress`

dekobon commented:
> > I just noticed that `EncryptedServerSideMultipartManagerIT` doesn't have `canListUploadsInProgress` marked as `@Test(groups = { "expensive" })`.
> 
> The [discussion](https://github.com/joyent/java-manta/pull/492#issuecomment-475365013) here didn't include `EncryptedServerSideMultipartManagerIT.canListUploadsInProgress`

Yeah, that omission is my fault. :(

nairashwin952013 commented:
> Glad to hear I'm not crazy about the counts. Per our call, we have two use cases we want to be easy to from the command line (for jenkins). Anything fancier is okay as a convenience for developers and their IDEs:
> (1) Run all of the tests we care about. This would be for nightly and after a commit to master.
> (2) Run all of the tests we care about, _except_ the really expensive ones. This would be for testing PRs.
> 
> With this PR the answer to (1) is just the default from running `mvn verify`. What is answer for (2)?

After discussing with Elijah, reached the conclusion that **all** `expensive tests` will be excluded from `mvn verify` which will be an answer to both aforementioned use-cases.

nairashwin952013 commented:
Output for `mvn clean verify -Dit.dryRun=true`
```
[main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 249
[main] WARN  com.joyent.test.util.MantaPathSuiteListener [ ] - Skipping suite cleanup since dry-run is enabled.
```

nairashwin952013 commented:
Pertinent to this PR, I wanted to state that if we are going to exclude the `expensive` testing group from `mvn verify` then I removed that test definition from the integration-test suite `testng-it.xml`. 

But if we want to invoke and run `expensive` using `-Dgroups` functionality then I will have to include the heavyweight definition which will mean to exclude `expensive` using `-DexcludedGroups=expensive` testing group.

dekobon commented on java-manta-it/src/test/java/com/joyent/test/util/TestListingInterceptor.java:
```diff
@@ -0,0 +1,89 @@
+            return;
+        }
+
+        final String[] sortedTests = observedTests.toArray(new String[0]);
```
This complicated conversion to an array can be avoided by using Java's streaming API:

```java
        if (LOG.isInfoEnabled()) {
            LOG.info("DRY-RUN: Listing [{}] tests that would have run",
                    observedTests.size());
            try (Stream<String> stream = observedTests.stream()) {
                stream.sorted().forEach(LOG::info);
            }
        }
```

nairashwin952013 replied:
Makes Sense.

dekobon commented on java-manta-it/src/test/java/com/joyent/test/util/TestListingInterceptor.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Override
+    public List<IMethodInstance> intercept(final List<IMethodInstance> methods, final ITestContext context) {
```
Please add a comment to describe what you are intending to do here.

nairashwin952013 replied:
`javadoc` Added.

dekobon commented on java-manta-it/src/test/java/com/joyent/test/util/TestListingInterceptor.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * Test method listener for determining which tests will actually run.
```
Please add a comment about why you would even want this test listener in the first place.

nairashwin952013 replied:
Done.

russelldb commented:
If I run `mvn verify it.dryRun=true` as per the PR initial description I get 
```
[ERROR] Unknown lifecycle phase "it.dryRun=true". You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]
```
After a bit more reading, it seems you mean `mvn verify -Dit.dryRun=true` (it's correct further into your initial PR description)

All that remains is for me to ask why you opened this PR? What problem you solve, and how I can determine the problem has been solved (or feature added)?

dekobon commented on java-manta-it/src/test/resources/testng-it.xml:
```diff
@@ -11,78 +12,87 @@
+            <class name="com.joyent.manta.client.MantaClientIT" />
+            <class name="com.joyent.manta.client.MantaClientPutIT" />
+            <class name="com.joyent.manta.client.MantaClientMetadataIT" />
+            <class name="com.joyent.manta.client.MantaObjectOutputStreamIT" />
```
You may want to comment out the line `<class name="com.joyent.manta.client.MantaObjectOutputStreamIT" />` for now because this test will fail due to #530.

nairashwin952013 replied:
Done.

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/test/util/TestListingInterceptor.java:
```diff
@@ -0,0 +1,89 @@
+        final String[] sortedTests = observedTests.toArray(new String[0]);
+        Arrays.sort(sortedTests);
+
+        LOG.info("DRY-RUN: Listing ["+ sortedTests.length +"] tests that would have run");
```
Done.

nairashwin952013 commented on java-manta-it/src/test/resources/testng-it.xml:
```diff
@@ -11,78 +12,87 @@
+            <class name="com.joyent.manta.client.MantaDirectoryListingIteratorIT" />
+        </classes>
+    </test>
+    <test name="Manta Client Http Headers Tests">
```
Done. assigned it under `<test name ="Manta Client Timeout Tests" >`

nairashwin952013 commented on java-manta-it/src/test/resources/testng-it.xml:
```diff
@@ -11,78 +12,87 @@
+	        -->
+        <classes>
+            <class name="com.joyent.manta.client.MantaClientFindIT" />
+            <class name="com.joyent.manta.client.MantaClientSigningIT" />
```
Removed.

nairashwin952013 commented on java-manta-it/src/test/resources/testng-it.xml:
```diff
@@ -3,6 +3,7 @@
 
```
Done.

nairashwin952013 commented:
When running `mvn verify` against the root directory.
```
[main] INFO  com.joyent.test.util.MantaPathSuiteListener [165.225.164.10 1b94e3c8-8967-11e9-b295-e336595f6b9e] - Removing base suite path: /ashwin.nair/stor/java-manta-integration-tests/$SUITE-UUID/
[INFO] Results:
[INFO] 
[WARNING] Tests run: 253, Failures: 0, Errors: 0, Skipped: 11



Total time: 24:37 min
```

nairashwin952013 commented on TESTING.md:
````diff
@@ -54,16 +54,35 @@ property or `MANTA_IT_NO_CLEANUP` environment variable.
+```
+
+Note: Since we have transitioned from wildcard to explicit test definition, it is **essential**
+for engineers to add tests complying with the nature of their corresponding test-groups.
````
Example given with the latest [commit](https://github.com/joyent/java-manta/pull/507/commits/44301aaaf07213c745a606a39be2c8b0658b4065)

nairashwin952013 commented:
> If I run `mvn verify it.dryRun=true` as per the PR initial description I get
> 
> ```
> [ERROR] Unknown lifecycle phase "it.dryRun=true". You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]
> ```
> 
> After a bit more reading, it seems you mean `mvn verify -Dit.dryRun=true` (it's correct further into your initial PR description)
> 
> All that remains is for me to ask why you opened this PR? What problem you solve, and how I can determine the problem has been solved (or feature added)?

**Goals & Challenges** updated in the PR Description along with additional changes in javadocs and testing documentation.

dekobon commented on TESTING.md:
````diff
@@ -58,19 +58,30 @@ Run `mvn verify` from the project root to run all tests. Some Maven goals will n
+`-Dtestnames=Manta Client Error Tests,Manta Client Timeout Tests`, e.g
 ```
-mvn test -DexcludedGroups=unlimited-crypto
+mvn test -Dtestnames=Manta Client Error Tests,Manta Client Timeout Tests
````
This example doesn't work. Did you test it?

nairashwin952013 replied:
It seems [Test-NG Docs](https://testng.org/doc/documentation-main.html#running-testng) haven't been updated as to the fact that Test-Ng runner parameters for `-testname`, `-testnames` aren't supported anymore, should've verified the changes before checking them in, in my perspective it will be wise to **revert** back to using `-groups` as *test runner params* .

dekobon commented on TESTING.md:
````diff
@@ -58,19 +58,30 @@ Run `mvn verify` from the project root to run all tests. Some Maven goals will n
+
 ```
-mvn clean verify -Dgroups=headers
+mvn clean verify -Dtestname=Manta Client Snaplinks Tests
````
This example also doesn't work.

cburroughs commented:
> > Glad to hear I'm not crazy about the counts. Per our call, we have two use cases we want to be easy to from the command line (for jenkins). Anything fancier is okay as a convenience for developers and their IDEs:
> > (1) Run all of the tests we care about. This would be for nightly and after a commit to master.
> > (2) Run all of the tests we care about, _except_ the really expensive ones. This would be for testing PRs.
> > With this PR the answer to (1) is just the default from running `mvn verify`. What is answer for (2)?
> 
> After discussing with Elijah, reached the conclusion that **all** `expensive tests` will be excluded from `mvn verify` which will be an answer to both aforementioned use-cases.
> 
> This essentially means for following use-cases:
> 
>     1. execute `mvn clean verify -DincludedGroups=expensive`. if we are opting to test heavyweight operations involving `mpu` and `jobs`.
> 
>     2. execute `mvn clean verify`

I'm afraid I'm still having trouble getting the two uses case to work, this gives me 250 tests:

```
$ mvn --projects java-manta-it --also-make  verify -Dit.dryRun=true | grep 'Expected test count'
[main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 250
```

But neither of these are giving me > 250 tests

```
 $ mvn --projects java-manta-it --also-make  verify -DincludedGroups=expensive -Dit.dryRun=true | grep 'Expected test count'
[main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 250
 $ mvn --projects java-manta-it --also-make verify -Dgroups=expensive -Dit.dryRun=true | grep 'Expected test count'
[main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 0
```
What is the set of flags to run all of the tests as we would for post-commit CI?

nairashwin952013 commented:
So we are entirely removing the approach of having an `expensive` test-group due to `TestNG` being extremely unreliable in its framework to support testng group exclusions cleanly as it was mentioned in #364 as well.  

All integration-tests that have been anointed as `expensive` have been removed from the default Testing suite file so all we have to do now is to run `mvn verify` as a Jenkins job for all pending PR's since the default test-suite is already running a subset of all the integration-tests (excluding the unimportant:`MantaClientJobIT`, `EncryptedJobsMultipartManager` etc. and deprecated integration-tests).

nairashwin952013 commented:
> > > Glad to hear I'm not crazy about the counts. Per our call, we have two use cases we want to be easy to from the command line (for jenkins). Anything fancier is okay as a convenience for developers and their IDEs:
> > > (1) Run all of the tests we care about. This would be for nightly and after a commit to master.
> > > (2) Run all of the tests we care about, _except_ the really expensive ones. This would be for testing PRs.
> > > With this PR the answer to (1) is just the default from running `mvn verify`. What is answer for (2)?
> > 
> > 
> > After discussing with Elijah, reached the conclusion that **all** `expensive tests` will be excluded from `mvn verify` which will be an answer to both aforementioned use-cases.
> > This essentially means for following use-cases:
> > ```
> > 1. execute `mvn clean verify -DincludedGroups=expensive`. if we are opting to test heavyweight operations involving `mpu` and `jobs`.
> > 
> > 2. execute `mvn clean verify`
> > ```
> 
> I'm afraid I'm still having trouble getting the two uses case to work, this gives me 250 tests:
> 
> ```
> $ mvn --projects java-manta-it --also-make  verify -Dit.dryRun=true | grep 'Expected test count'
> [main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 250
> ```
> 
> But neither of these are giving me > 250 tests
> 
> ```
>  $ mvn --projects java-manta-it --also-make  verify -DincludedGroups=expensive -Dit.dryRun=true | grep 'Expected test count'
> [main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 250
>  $ mvn --projects java-manta-it --also-make verify -Dgroups=expensive -Dit.dryRun=true | grep 'Expected test count'
> [main] INFO  com.joyent.test.util.MantaPathSuiteListener [ ] - Expected test count: 0
> ```
> 
> What is the set of flags to run all of the tests as we would for post-commit CI?

We would just be running these tests like we run normal integration-tests. For instance,
`mvn verify -Dtest=MantaClientJobtIT`

cburroughs commented:
In that case, is it still necessary to switch  from wildcards to a explicit list in the testng suite file?  I thought we did that as part of trying to get groups to work, but it has an ongoing cost (future engineers will inevitably mess it up)

nairashwin952013 commented:
> In that case, is it still necessary to switch from wildcards to a explicit list in the testng suite file? I thought we did that as part of trying to get groups to work, but it has an ongoing cost (future engineers will inevitably mess it up)

Yes but I explicitly added that in the documentation [here](https://github.com/joyent/java-manta/pull/507/files#diff-84889d0af78b35050da8d12673769ebaR74) that developers will have to add definitions to the testing suite once they come up with a new integration-test. I will make sure it gets added to the general SDK User Guide as well. Also, class-level definitions were in-place to reduce the redundancies significantly otherwise were running each integration-test almost 6 times probably more.

nairashwin952013 commented on TESTING.md:
````diff
@@ -54,16 +54,50 @@ property or `MANTA_IT_NO_CLEANUP` environment variable.
+mvn clean verify -Dgroups=headers
+```
+
+Note: Since we have transitioned from wildcard to explicit test definition, it is **essential**
````
The tests will run nevertheless of being explicitly enumerated **but** they will not run if we source the `testing-it.xml` file since we have transitioned from a `package-level` -> `class-level` hierarchy. I will make sure to add this to `README` and incorporate the edit

nairashwin952013 commented on TESTING.md:
````diff
@@ -54,16 +54,50 @@ property or `MANTA_IT_NO_CLEANUP` environment variable.
+mvn verify -Dit.dryRun=true
+```
+	
+Note: This has not yet been expanded to unit tests.
````
You're absolutely correct this won't be incorporated into unit-tests. I merely stated it so that the SDK user doesn't erroneously assume that `dry-Run` is available as a feature for unit-tests too.
Maybe this sentence should be re-edited from
`Note: This has not yet been expanded to unit tests.` -> `Note: This feature is invalid for unit tests.`

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/test/util/TestListingInterceptor.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Test method listener for determining which tests will actually run
+ * when we invoke mvn verify -Dit.dryRun=true. This test listener has
+ * been introduced to leverage TestNG feature @since 6.8.21 of determining
```
Noted.

nairashwin952013 commented on java-manta-it/src/test/java/com/joyent/manta/config/IntegrationTestConfigContext.java:
```diff
@@ -52,6 +53,16 @@ public IntegrationTestConfigContext(Boolean usingEncryption) {
+     */
+    public IntegrationTestConfigContext(final String encryptionCipher) {
+        this(encryptionCipher != null, encryptionCipher);
+
```
Removed.

nairashwin952013 commented:
> I noted a few minor doc or formatting nits on the last read through. This is good to land once those are folded in. Thanks for shepherding this big set of changes through.

 Thanks @cburroughs for this thorough review, learned a lot of much needed conventions, basics through this entire process.

dekobon commented:
I ran the integration tests for the latest bits and found the following test failure:

```
[ERROR] Tests run: 256, Failures: 1, Errors: 0, Skipped: 14, Time elapsed: 953.694 s <<< FAILURE! - in TestSuite
[ERROR] pruneParentDirectoriesFull(com.joyent.manta.client.MantaClientDirectoriesIT)  Time elapsed: 29.197 s  <<< FAILURE!
java.lang.AssertionError: expected [false] but found [true]
        at org.testng.Assert.fail(Assert.java:96)
        at org.testng.Assert.failNotEquals(Assert.java:776)
        at org.testng.Assert.assertFalse(Assert.java:65)
        at org.testng.Assert.assertFalse(Assert.java:75)
        at com.joyent.manta.client.MantaClientDirectoriesIT.pruneParentDirectoriesFull(MantaClientDirectoriesIT.java:262)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:583)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:719)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:989)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

nairashwin952013 commented:
> I ran the integration tests for the latest bits and found the following test failure:
> 
> ```
> [ERROR] Tests run: 256, Failures: 1, Errors: 0, Skipped: 14, Time elapsed: 953.694 s <<< FAILURE! - in TestSuite
> [ERROR] pruneParentDirectoriesFull(com.joyent.manta.client.MantaClientDirectoriesIT)  Time elapsed: 29.197 s  <<< FAILURE!
> java.lang.AssertionError: expected [false] but found [true]
>         at org.testng.Assert.fail(Assert.java:96)
>         at org.testng.Assert.failNotEquals(Assert.java:776)
>         at org.testng.Assert.assertFalse(Assert.java:65)
>         at org.testng.Assert.assertFalse(Assert.java:75)
>         at com.joyent.manta.client.MantaClientDirectoriesIT.pruneParentDirectoriesFull(MantaClientDirectoriesIT.java:262)
>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>         at java.lang.reflect.Method.invoke(Method.java:498)
>         at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
>         at org.testng.internal.Invoker.invokeMethod(Invoker.java:583)
>         at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:719)
>         at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:989)
>         at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
>         at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
>         at java.lang.Thread.run(Thread.java:748)
> ```

It seems this particular failure occurred due to the `STRING_GENERATOR` making some discrepancies when it came to creating the `parentDir` and `childDir` directories and some inconsistencies with their `depths`. 
I ran this specific test separately and ran it while I invoked `mvn verify` and I didn't encounter any failures perhaps you could attach both` String` values so that I could interpret as to the reason it failed but the changes that have been checked in within #507 has nothing to do with this failure.

dekobon commented:
I'm going to go ahead and approve the PR. Can you make another ticket to capture the problems with that particular test?

nairashwin952013 commented:
> I'm going to go ahead and approve the PR. Can you make another ticket to capture the problems with that particular test?

Yes. Noted

nairashwin952013 commented:
This branch has been squashed and merged into `master`.
